code/game/objects/items/weapons/weapon.dm

```diff
@@ -10,3 +9,178 @@
 	var/max_shots = 0 //codex, energy weapons
 	var/scope_zoom = FALSE//codex
 	var/self_recharge = FALSE //codex
+
+	///Bayonet Charge base obj
+	var/datum/action/bayonet_charge/charge
+//its here because if I dont place it in the same script as the base weapon define, it breaks.
+/datum/action/bayonet_charge
+	name = "Toggle Bayonet Charge"
+	action_icon_state = "charge_human"
+
+	//these variables modify the stats of a bayonet charge
+	///the amount of tiles you need to walk to start charging
+	var/minactivationrundist = 5
+	///thhe amount of tiles you need to walk to charge fully
+	var/fullspeedrundist = 9
+	///speed increase per step once charging
+	var/speedperstep = 0.2
+	///MAXIMUMSPEED INCREASE
+	var/maxspeed = 2.4
+	///the amount by which the stam loss is multiplied by per step taken when charging
+	var/stamlossmult = 1
+	///the multiplier for damage inflicted when you bayonet charge someone
+	var/damagemult = 1.5
+	///This is a speed mult for when you finish charging
+	var/postspeedmult = 1.5
+	///the amount of time the post charge speed boost lasts (in deciseconds)
+	var/postspeedtime = 50
+	///the amount of time the stun lasts in 1/10ths of a second
+	var/slownesstime = 5
+	///the amount of time cooldown lasts in deciseconds
+	var/cooldowntime = 150
+
+	//Below are static variables that should not be changed, ever.
+	var/stepstaken = 0
+	var/chargedir = null
+
+	var/amcharging = FALSE
+	var/chargewhenmove = FALSE
+	var/incooldown = FALSE
+	var/obj/item/weapon/weaponinhand
+
+	///represents the time when the stepstaken var resets
+	var/movementcontinuationlimit
+	///the guy that is charging
+	var/mob/living/carbon/human/charger
+
+	///The list of melee/not gun weapons available to be charged with
+	var/list/enabled_melee_weapons = list(
+		/obj/item/weapon/powerfist,
+		/obj/item/weapon/claymore/mercsword/machete,
+		/obj/item/weapon/claymore/mercsword/officersword,
+		/obj/item/weapon/twohanded/spear/tactical,
+	)
+
+/obj/item/weapon/Initialize()
+	. = ..()
+	charge = new
+
+/obj/item/weapon/dropped(mob/user)
+	. = ..()
+	charge.remove_action(user)
+
+datum/action/bayonet_charge/give_action(mob/M, obj/item/weapon/W)
+	. = ..()
+	weaponinhand = W
+	charger = owner
+
+datum/action/bayonet_charge/remove_action(mob/M)
+
+	charger = owner
+	charge_off()
+	. = ..()
+
+datum/action/bayonet_charge/action_activate()
+	if(amcharging)
+		return
+	if(chargewhenmove == TRUE)
+		charge_off()
+		return
+	charger = owner
+	charge_on()
+datum/action/bayonet_charge/proc/charge_on()
+	if(incooldown)
+		to_chat(charger, "<span class='warning'>Charge is still on cooldown!</span>")
+		return
+	chargewhenmove = TRUE
+	src.add_selected_frame()
```

Comment:
src.var or src.proc() is stricly equivalent to var and proc().

Do not use src.something